flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java

```diff
@@ -123,6 +127,15 @@ public void submitSessionCluster(FlinkDeployment deployment, Configuration conf)
                                     .get(1, TimeUnit.MINUTES);
                     savepointOpt = Optional.of(savepoint);
                     break;
+                case LAST_STATE:
+                    if (!HighAvailabilityMode.isHighAvailabilityModeActivated(conf)) {
+                        throw new InvalidDeploymentException(
```

Comment:
This check/validation can be done in advance, not in the `cancel` phase. We may move this to the [shared validation logic](https://issues.apache.org/jira/browse/FLINK-26136) later ?

Reply:
Good point, let's keep it here I will open a PR soon with the extracted validation logic. I will move it there once this PR is merged

Reply:
Maybe we need some documentation on how this should work eventually? There could be a recent checkpoint, even when HA is not enabled. That could be queried through the REST API. Though that path hinges on the availability of the REST API, it would still allow for the upgrade.

Reply:
The rest API is much trickier as there is no guarantee that it is available once the job is stopped if something happens to the process. I think it is a fair requirement that last_state only works with ha enabled now, and maybe we can improve this in the long run.

Reply:
I think for production use-cases this won't be a problem anyways

Reply:
Yes. The pre-check could be done in webhook validation logic.

I agree with @gyfora that relying on REST API is not stable enough. Maybe FLINK-24113 is useful in such case. Since enabling the K8s HA for Flink cluster does not bring any burden, I think most of the users is running Flink applications with HA enabled in production environment. Let's support last-state upgrade without HA in the future if necessary.